README.md

```diff
@@ -1 +1,46 @@
 # Alc the Ripper
+Alc the Ripper (Alc for short) is a *state-of-the-art* cloud-based password cracker.
+The user submits a salt, a PBKDF2 digest, and the length of the password, then Alc will use brute force to find the password.
+
+# Design Requirements
+* Automation - Alc should not require any human intervention when running. The user simply submits a request and Alc will do its job and then report the finding.
+* Elasticity - Alc should handle variable user demands and offer the same level of service (in terms of hashes per second) to all the users.
+* Load Balancing - The jobs should be evenly spread out across all the slaves to achieve maximum performance.
+* Reliability - Jobs should checkpointed and restarted from the checkpoint when failures occur.
+* Monitoring - Maintain metrics about the whole system to monitor job status, resource usage and so on.
+* Scheduling - TODO
+* Multi-tenancy - TODO
+* Security - TODO
+
+# Design Overview
+Alc is designed to run on IaaS providers such as AWS.
+It uses the master-slave model. Where the master assign tasks to the slaves and the slaves periodically send heartbeat messages and status updates back to the master.
+
+We achieve elasticity using the [PID controller](https://en.wikipedia.org/wiki/PID_controller).
+Before getting into its functionality, we first define the meaning of resources in Alc.
+The number of resources is the number of tasks that can be carried out in parallel in a time instant, somewhat analogous to [DOP](https://en.wikipedia.org/wiki/Degree_of_parallelism).
+For example, if an EC2 instance can run two tasks in parallel, then that instance represents two units of resource.
+For the PID controller, we define the error as `Ra - Rr`, where `Ra` is the total available resource, and `Rr` is the total required resource.
```

Comment:
Total available resource is equal to how much we can pay AWS. Therefore, I think "total available" is not correctly phrased.